modular_skyrat/modules/mob_container/mob_container.dm

```diff
@@ -0,0 +1,177 @@
+
+/**
+ * Used to check whether or not a mob has a container. We don't want to contain mobs that don't have containers.
+ */
+/datum/component/mob_container/proc/container_exist_check()
+	return TRUE
+
+/**
+ * Proc to leave the container we're currently inside of.
+ */
+/datum/component/mob_container/proc/leave_container(datum/source)
+	SIGNAL_HANDLER
+
+	var/mob/our_container = mob_holder.loc
+
+	if(!istype(our_container))
+		return
+
+	SEND_SIGNAL(our_container, COMSIG_CONTAINER_REMOVE_MOB, mob_holder)
+
+/**
+ * Signal handler for storing a mob.
+ */
+/datum/component/mob_container/proc/store_mob_handler(datum/source, mob/to_store)
+	SIGNAL_HANDLER
+
+	if(!to_store)
+		return
+
+	store_mob_explicitly(to_store)
+
+/**
+ * Signal handler for removing a mob.
+ */
+/datum/component/mob_container/proc/remove_mob_handler(datum/source, mob/to_remove)
+	SIGNAL_HANDLER
+
+	if(!to_remove)
+		return
+
+	remove_mob_explicitly(to_remove)
+
+/**
+ * Signal handler for dumping every mob inside of us.
+ */
+/datum/component/mob_container/proc/dump_all_mobs()
+	SIGNAL_HANDLER
+
+	for(var/mob/contained in mobs_held)
+		remove_mob_explicitly(contained)
+
+/**
+ * Stores a mob inside of our parent mob.
+ *
+ * mob/to_store | the mob to store inside of us
+ * force | if TRUE, we'll store a containerless mob
+ */
+/datum/component/mob_container/proc/store_mob_explicitly(mob/to_store, force = FALSE)
+	if(!istype(to_store))
+		return
+
+	if(!SEND_SIGNAL(to_store, COMSIG_MOB_HAS_CONTAINER) && !force)
+		return
+
+	if(!SEND_SIGNAL(mob_holder, COMSIG_MOB_CONTAINER_ENTERED, to_store) || !SEND_SIGNAL(to_store, COMSIG_MOB_ENTER_CONTAINER, mob_holder))
+		return
+
+	to_store.forceMove(mob_holder)
+	to_store.overlay_fullscreen("in_mob_container", fullscreen_overlay, 1)
+	mobs_held |= to_store
+
+/**
+ * Removes a mob from inside our parent mob.
+ *
+ * mob/to_store | the mob to remove from us
+ */
+/datum/component/mob_container/proc/remove_mob_explicitly(mob/to_remove)
+	if(!istype(to_remove) && !(to_remove in mob_holder || to_remove.loc == mob_holder))
+		return
+
+	if(!SEND_SIGNAL(mob_holder, COMSIG_MOB_CONTAINER_EXITED, to_remove) || !SEND_SIGNAL(to_remove, COMSIG_MOB_EXIT_CONTAINER, mob_holder))
+		return
+
+	to_remove.forceMove(get_turf(mob_holder))
+	to_remove.clear_fullscreen("in_mob_container")
+	mobs_held -= to_remove
+
+/**
+ * Called whenever we enter another mob. If our return value is FALSE, we won't enter the mob.
+ *
+ * mob/mob_were_entering | the mob we're being put inside of
+ */
+/datum/component/mob_container/proc/on_enter_container(datum/source, mob/mob_were_entering)
+	SIGNAL_HANDLER
+
+	return TRUE
+
+/**
+ * Called whenever we exit another mob. If our return value is FALSE, we won't exit the mob.
+ *
+ * mob/mob_were_exiting | the mob we're being taken out of
+ */
+/datum/component/mob_container/proc/on_exit_container(datum/source, mob/mob_were_exiting)
```

Comment:
Any good reason this ought to be called by sending a COMSIG_MOB_CONTAINER_EXITED signal rather than doing so right away on line 132? Don't get me wrong, the signals are ok, but calling "native" procs this way is pretty convoluted, inefficient.
Also, since you're adding so signals already, perhaps you should ditch the idea of having subtypes of this component with their own quirks. Let it be done by other datums, components or elements and keep it simple here.

Reply:
Meh, it isn't a huge deal. Efficiency wise, it isn't too terribly different from calling it natively (unless the load is large), however I think this is something worth changing.

Component subtypes aren't a new thing by any means. I'm going to leave it on the subtype idea, that can be changed in the future if absolutely dire.

Reply:
Meh, it isn't a huge deal. Efficiency wise, it isn't too terribly different from calling it natively (unless the load is large), however I think this is something worth changing.

Component subtypes aren't a new thing by any means. I'm going to leave it on the subtype idea, that can be changed in the future if absolutely dire.